lib/core/src/Cardano/Pool/Metrics.hs

```diff
@@ -274,20 +276,24 @@ newStakePoolLayer db@DBLayer{..} nl tr = StakePoolLayer
 
         (distr, prod) <- liftIO . atomically $ (,)
             <$> (Map.fromList <$> readStakeDistribution nodeEpoch)
-            <*> (count <$> readPoolProduction nodeEpoch)
+            <*> readPoolProduction nodeEpoch
+
+        let prodTip = maximumBy (compare `on` view #slotId)
```

Comment:
From the docs for [`maximumBy`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-List.html#v:maximumBy):
> The largest element of a non-empty structure with respect to the given comparison function.

If there are no pool productions in the epoch of the node tip, this will crash. I.e when it should have  thrown `ErrMetricsIsUnsynced`, it will crash instead.

You cannot move this binding to after the `when (Map.null distr || Map.null prod` check, since `computeProgress` relies on the `prodTip`.

I would suggest using `poolProductionTip`. You can put it inside the `atomically` and there shouldn't be any downsides:

```haskell
(distr, prod, dbTip) <- liftIO . atomically $ (,,)
             <$> (Map.fromList <$> readStakeDistribution nodeEpoch)
             <*> readPoolProduction nodeEpoch
             <*> (readPoolProductionCursor 1 >>= toMaybe)
```

Reply:
Hmmm. Good point!